model_hub/examples/huggingface/language-modeling/clm_trial.py

```diff
@@ -0,0 +1,163 @@
+        self.tokenized_datasets = self.build_datasets()
+        train_length = len(self.tokenized_datasets["train"])
+        self.logger.info("training records: {}".format(train_length))
+        if (
+            "records_per_epoch" in self.exp_config
+            and train_length != self.exp_config["records_per_epoch"]
+        ):
+            self.logger.warning(
+                "number of train records {} does not match records_per_epoch of {}".format(
+                    train_length, self.exp_config["records_per_epoch"]
+                )
+            )
+
+        self.reducer = self.context.experimental.wrap_reducer(
+            lambda losses: np.exp(np.mean(losses)), name="perplexity", for_training=False
+        )
+
+    def build_datasets(self) -> Union[datasets.Dataset, datasets.DatasetDict]:
+        column_names = self.raw_datasets["train"].column_names
+        text_column_name = "text" if "text" in column_names else column_names[0]
+
+        def tokenize_function(tokenizer, examples):
+            return tokenizer(examples[text_column_name])
+
+        # We cannot use self.tokenizer as a non-local variable in the tokenize_function if we want
+        # map to be able to cache the output of the tokenizer.  Hence, the tokenize_function takes
+        # a tokenizer explicitly as an input and we create a closure using functools.partial.
+        tokenized_datasets = self.raw_datasets.map(
+            functools.partial(tokenize_function, self.tokenizer),
+            batched=True,
+            num_proc=self.data_config.preprocessing_num_workers,
+            remove_columns=column_names,
+            load_from_cache_file=not self.data_config.overwrite_cache,
+        )
+
+        if self.data_config.max_seq_length is None:
+            max_seq_length = self.tokenizer.model_max_length
+            if max_seq_length > 1024:
+                self.logger.warning(
+                    "The tokenizer picked seems to have a very large `model_max_length` "
+                    f"({self.tokenizer.model_max_length}). Using 1024 instead. You can change "
+                    "that default value by setting max_seq_length in the experiment config."
+                )
+            max_seq_length = 1024
+        else:
+            if self.data_config.max_seq_length > self.tokenizer.model_max_length:
+                self.logger.warning(
+                    f"The max_seq_length passed ({self.data_config.max_seq_length}) is larger "
+                    f"than the maximum length for the model ({self.tokenizer.model_max_length}). "
+                    f"Using max_seq_length={self.tokenizer.model_max_length}."
+                )
+            max_seq_length = min(self.data_config.max_seq_length, self.tokenizer.model_max_length)
+
+        # Main data processing function that will concatenate all texts from our dataset and
+        # generate chunks of max_seq_length.
+        def group_texts(examples):
```

Comment:
I understand that the code right below this uses the nested function here, but not sure structure-wise if it is cleaner to put this right under `build_datasets` or preferably to separate it out entirely to a data.py file? Ditto this for mlm and plm.

Reply:
unless you feel strongly, i prefer to leave it as is since this closely follows the structure of the original hf example